### ISSUE TITLE
Exoplayer not starting playback after downloading DEFAULT_BUFFER_FOR_PLAYBACK_MS of data

### ISSUE THREAD

**Issue by krackjack234** [google/ExoPlayer#10070]
Question
======

We are trying to start MPD playback as soon as DEFAULT_BUFFER_FOR_PLAYBACK_MS of data is downloaded.
In our case, each mp4 segment is of 6s which means playback should start after downloading of 1 video segment as we have configured DEFAULT_BUFFER_FOR_PLAYBACK_MS as 1000ms

However, we see that the first frame is getting rendered after downloading 3-4 video segments even at a resolution of 360p

Below are the logs that we printed

```
2022-03-14 22:16:16.899 31112-31112/com.app.exoplayersamplecomponent D/Profiling::: Preparing Player
2022-03-14 22:16:17.667 31112-31112/com.app.exoplayersamplecomponent E/Profiling::: bitrate: 100000 bytes loaded: 10084
2022-03-14 22:16:17.969 31112-31112/com.app.exoplayersamplecomponent E/Profiling::: bitrate: 100000 bytes loaded: 91080
2022-03-14 22:16:18.130 31112-31112/com.app.exoplayersamplecomponent E/Profiling::: bitrate: 100000 bytes loaded: 127478
2022-03-14 22:16:18.221 31112-31112/com.app.exoplayersamplecomponent E/Profiling::: bitrate: 100000 bytes loaded: 78050
2022-03-14 22:16:18.257 31112-31112/com.app.exoplayersamplecomponent E/Profiling::: bitrate: 2300000 bytes loaded: 10090
2022-03-14 22:16:18.653 31112-31112/com.app.exoplayersamplecomponent E/Profiling::: First Frame Rendered
```

We have set initial bitrate to 100kbps and player is configured in ABR mode

We printed the bytes downloaded in **onLoadCompleted** callback for only video segments

As we see in above logs, the first frame rendered callback is coming ONLY after downloading of 4 segments which is delaying the start of playback.

Please help to understand this behavior. We want the player to start as soon as  DEFAULT_BUFFER_FOR_PLAYBACK_MS is downloaded which is not happening now.

Our Exoplayer version is 2.14.1

**Comment by icbaker** [comment]
I'm not sure the logs you've provided conclusively show that the library is **waiting** for 3-4 segments before starting playback. Segments will continue to be loaded in parallel while playback starts, so if your network connection is fast and your segments are small it's reasonable that many of them are downloaded and buffered while the player is starting to play the first segment (and before it's managed to render the first frame).

What you really want to see is the point the player switches from 'buffering' to 'ready' state (available through [`AnalyticsListener#onPlaybackStateChanged`](https://exoplayer.dev/doc/reference/com/google/android/exoplayer2/analytics/AnalyticsListener.html#onPlaybackStateChanged(com.google.android.exoplayer2.analytics.AnalyticsListener.EventTime,@com.google.android.exoplayer2.Player.Stateint))). In fact, there's lots of interesting logging available if you just wire up an [`EventLogger`](https://exoplayer.dev/debug-logging.html), because that will also include decoder initialization etc. Can you try wiring that up and replying with the more detailed logging?

> Our Exoplayer version is 2.14.1

Do you see the same behaviour if you make the `DEFAULT_BUFFER_FOR_PLAYBACK_MS` change in the 2.17.1 [ExoPlayer demo app](https://exoplayer.dev/demo-application.html)? It should be relatively easy to try.

What device are you seeing this behaviour on?

**Comment by krackjack234** [comment]
Hi @icbaker  I think we atleast figured out why the player was downloading 3-4 segments before first frame was getting rendered.

We used the below code to initialise player.

```
  DefaultBandwidthMeter bandwidthMeter = new DefaultBandwidthMeter.Builder(context)
                        .setInitialBitrateEstimate(initialBitrate)
                        .build();

  player = new SimpleExoPlayer.Builder(context, renderersFactory)
                        .setMediaSourceFactory(mediaSourceFactory)
                        .setTrackSelector(trackSelector)
                        .setLoadControl(loadControl)
                       .setBandwidthMeter(bandwidthMeter)
```
If we comment out the bandwidthMeter, the player then starts rendering just after downloading first frame which was our desired behavior.

Pls let us know if this observation makes sense.

**Comment by krackjack234** [comment]
@icbaker  Here are one more log showing when the player is going to ready state. This is with code where we are calling 
setInitialBitrateEstimate and adding the bandwidth meter

```
2022-03-15 00:37:43.473 11078-11078/com.app.exoplayersamplecomponent D/Profiling::: creating Player
2022-03-15 00:37:43.497 11078-11078/com.app.exoplayersamplecomponent E/Profiling::: state: 1 playWhenReady: true
2022-03-15 00:37:43.498 11078-11078/com.app.exoplayersamplecomponent D/Profiling::: Player created
2022-03-15 00:37:43.505 11078-11078/com.app.exoplayersamplecomponent D/Profiling::: Preparing Player
2022-03-15 00:37:43.741 11078-11078/com.app.exoplayersamplecomponent E/Profiling::: state: 2 playWhenReady: true
2022-03-15 00:37:44.230 11078-11078/com.app.exoplayersamplecomponent E/Profiling::: bitrate: 100000 bytes loaded: 10084
2022-03-15 00:37:44.440 11078-11078/com.app.exoplayersamplecomponent E/Profiling::: bitrate: 100000 bytes loaded: 91080
2022-03-15 00:37:44.827 11078-11078/com.app.exoplayersamplecomponent E/Profiling::: bitrate: 100000 bytes loaded: 127478
2022-03-15 00:37:45.040 11078-11078/com.app.exoplayersamplecomponent E/Profiling::: bitrate: 100000 bytes loaded: 78050
2022-03-15 00:37:45.126 11078-11078/com.app.exoplayersamplecomponent E/Profiling::: First Frame Rendered
2022-03-15 00:37:45.192 11078-11078/com.app.exoplayersamplecomponent D/Profiling::: Player Ready
2022-03-15 00:37:45.192 11078-11078/com.app.exoplayersamplecomponent E/Profiling::: state: 3 playWhenReady: true
2022-03-15 00:37:45.213 11078-11078/com.app.exoplayersamplecomponent E/Profiling::: bitrate: 2300000 bytes loaded: 10090
2022-03-15 00:37:45.831 11078-11078/com.app.exoplayersamplecomponent E/Profiling::: bitrate: 2300000 bytes loaded: 1428661
2022-03-15 00:37:46.993 11078-11078/com.app.exoplayersamplecomponent E/Profiling::: bitrate: 2300000 bytes loaded: 1971037
```

If we remove bandwidthMeter, then the player starts after downloading 1-2 segments as seen in below logs

```
2022-03-15 00:40:54.433 12475-12475/com.app.exoplayersamplecomponent D/Profiling::: creating Player
2022-03-15 00:40:54.449 12475-12475/com.app.exoplayersamplecomponent E/Profiling::: state: 1 playWhenReady: true
2022-03-15 00:40:54.450 12475-12475/com.app.exoplayersamplecomponent D/Profiling::: Player created
2022-03-15 00:40:54.453 12475-12475/com.app.exoplayersamplecomponent D/Profiling::: Preparing Player
2022-03-15 00:40:54.473 12475-12475/com.app.exoplayersamplecomponent E/Profiling::: state: 2 playWhenReady: true
2022-03-15 00:40:54.973 12475-12475/com.app.exoplayersamplecomponent E/Profiling::: bitrate: 2300000 bytes loaded: 10090
2022-03-15 00:40:56.030 12475-12475/com.app.exoplayersamplecomponent E/Profiling::: First Frame Rendered
2022-03-15 00:40:56.062 12475-12475/com.app.exoplayersamplecomponent D/Profiling::: Player Ready
2022-03-15 00:40:56.062 12475-12475/com.app.exoplayersamplecomponent E/Profiling::: state: 3 playWhenReady: true
2022-03-15 00:40:56.773 12475-12475/com.app.exoplayersamplecomponent E/Profiling::: bitrate: 2300000 bytes loaded: 2523959
2022-03-15 00:40:58.282 12475-12475/com.app.exoplayersamplecomponent E/Profiling::: bitrate: 2300000 bytes loaded: 2534093
```

**Comment by krackjack234** [comment]
Updating with decoder init logs using exoplayer v2.14.1. 

Without Bandwidth meter and setInitialBitrateEstimate call

```
2022-03-15 13:27:59.518 26073-26073/com.app.exoplayersamplecomponent D/Profiling::: creating Player
2022-03-15 13:27:59.533 26073-26073/com.app.exoplayersamplecomponent E/Profiling::: state: 1 playWhenReady: true
2022-03-15 13:27:59.534 26073-26073/com.app.exoplayersamplecomponent D/Profiling::: Player created
2022-03-15 13:27:59.538 26073-26073/com.app.exoplayersamplecomponent D/Profiling::: Preparing Player
2022-03-15 13:27:59.911 26073-26073/com.app.exoplayersamplecomponent E/Profiling::: state: 2 playWhenReady: true
2022-03-15 13:28:00.185 26073-26073/com.app.exoplayersamplecomponent E/Profiling::: bitrate: 2300000 bytes loaded: 10090
2022-03-15 13:28:00.454 26073-26073/com.app.exoplayersamplecomponent E/Profiling::: decoder init: trackType: 2 decoder: OMX.hisi.video.decoder.avc duration 0
2022-03-15 13:28:00.534 26073-26073/com.app.exoplayersamplecomponent E/Profiling::: decoder init: trackType: 1 decoder: c2.android.aac.decoder duration 0
2022-03-15 13:28:00.811 26073-26073/com.app.exoplayersamplecomponent E/Profiling::: bitrate: 2300000 bytes loaded: 2523959
2022-03-15 13:28:01.174 26073-26073/com.app.exoplayersamplecomponent E/Profiling::: First Frame Rendered
2022-03-15 13:28:01.199 26073-26073/com.app.exoplayersamplecomponent D/Profiling::: Player Ready
2022-03-15 13:28:01.199 26073-26073/com.app.exoplayersamplecomponent E/Profiling::: state: 3 playWhenReady: true
```

With BandwithEstimator and setInitialBitrateEstimate set to 100kbps

```
2022-03-15 13:34:44.715 29803-29803/com.app.exoplayersamplecomponent D/Profiling::: Preparing Player
2022-03-15 13:34:44.732 29803-29803/com.app.exoplayersamplecomponent E/Profiling::: state: 2 playWhenReady: true
2022-03-15 13:34:45.088 29803-29803/com.app.exoplayersamplecomponent E/Profiling::: bitrate: 100000 bytes loaded: 10084
2022-03-15 13:34:45.270 29803-29803/com.app.exoplayersamplecomponent E/Profiling::: decoder init: trackType: 2 decoder: OMX.hisi.video.decoder.avc duration 0
2022-03-15 13:34:45.273 29803-29803/com.app.exoplayersamplecomponent E/Profiling::: bitrate: 100000 bytes loaded: 91080
2022-03-15 13:34:45.317 29803-29803/com.app.exoplayersamplecomponent E/Profiling::: decoder init: trackType: 1 decoder: c2.android.aac.decoder duration 0
2022-03-15 13:34:45.323 29803-29803/com.app.exoplayersamplecomponent E/Profiling::: bitrate: 100000 bytes loaded: 127478
2022-03-15 13:34:45.359 29803-29803/com.app.exoplayersamplecomponent E/Profiling::: bitrate: 100000 bytes loaded: 78050
2022-03-15 13:34:45.394 29803-29803/com.app.exoplayersamplecomponent E/Profiling::: bitrate: 2300000 bytes loaded: 10090
2022-03-15 13:34:45.769 29803-29803/com.app.exoplayersamplecomponent E/Profiling::: bitrate: 2300000 bytes loaded: 1428661
2022-03-15 13:34:46.110 29803-29803/com.app.exoplayersamplecomponent E/Profiling::: First Frame Rendered
2022-03-15 13:34:46.243 29803-29803/com.app.exoplayersamplecomponent D/Profiling::: Player Ready
2022-03-15 13:34:46.243 29803-29803/com.app.exoplayersamplecomponent E/Profiling::: state: 3 playWhenReady: true
```

Unfortunately it will be difficult for us to move to latest version as there are some more dependent components that we have to update.

Please let us know your thoughts about this @icbaker

**Comment by krackjack234** [comment]
@icbaker  we analysed little more and found that **any** of the below code changes are causing the 2.14.1 exoplayer to delay rendering of first frame. We checked this in 2.14.1 exoplayer demo app

```
1.  DefaultBandwidthMeter bandwidthMeter = new DefaultBandwidthMeter.Builder(context)
                        .setInitialBitrateEstimate(logixPlayerBuilder.initialBitrate)
                        .build();
```

`2.  trackSelectorParameters =  
                         builder.setMaxVideoBitrate(1000).build();`

`3. trackSelectorParameters = 
                       builder.setMaxVideoSizeSd().build();`

If we donot use any of the above code, the first frame rendered event comes quickly after downloading only 1-2 video segments

Our goals are as follows

1. Set max resolution/bitrate to SD or any particular value for L3 devices as our streams have L1 tracks for 720p and higher resolutions
2. Start exoplayer with a lower bitrate and then go to the desired one based on bandwidth (for ABR) or as set by user (say 2mbps fixed bitrate)

Any thoughts?

We are seeing this issue on Xiaomi and Honor Mobiles but we think this should happen on all devices.

**Comment by krackjack234** [comment]
Hi @icbaker  

after further analysis, we figured out that the player is making call for drm license late (after download 3-4 video segments) which is causing the delay. Here is a snapshot from Charles


![Screen Shot 2022-03-19 at 5 57 10 PM](https://user-images.githubusercontent.com/50735078/159121663-77602c79-2237-4ec6-b2dd-9b87b7a97fcd.png)


Here the drm license call is made 2 secs after downloading the manifest( first call) and after downloading 4 video segments and 3 audio segments.

Is there any way we can ensure that the player is making the license call in advance while its downloading the segments?

It looks like the player is ultimately waiting for the license call to complete even when it has downloaded enough data to start playback.

**Comment by krackjack234** [comment]
hi @icbaker  any thoughts on how we can ensure that the player is making the license call as soon as it parses the manifest and does not wait for the initial segments to download?

**Comment by icbaker** [comment]
I don't see the same behaviour in the demo app when playing the `Widevine DASH (MP4, H264) > HD (cenc)` sample. I see the following (I added the logging to the `open()` method of `DefaultHttpDataSource` (used for DRM requests) and `CronetDataSource` (used for content).

With no `BandwidthMeter` customization:

```
2022-03-22 17:00:02.079 1950-2289/? W/cronet-open: https://storage.googleapis.com/wvmedia/cenc/h264/tears/tears.mpd
2022-03-22 17:00:02.139 1950-2295/? W/cronet-open: https://storage.googleapis.com/wvmedia/cenc/h264/tears/tears_h264_high_1080p_20000.mp4
2022-03-22 17:00:02.139 1950-2296/? W/cronet-open: https://storage.googleapis.com/wvmedia/cenc/h264/tears/tears_audio_eng.mp4
2022-03-22 17:00:02.297 1950-2296/? W/cronet-open: https://storage.googleapis.com/wvmedia/cenc/h264/tears/tears_audio_eng.mp4
2022-03-22 17:00:02.468 1950-2297/? W/dhds-open: https://proxy.uat.widevine.com/proxy?video_id=2015_tears&provider=widevine_test
```

With the `BandwidthMeter` customized with an `initialBitrateEstimate = 100_000`:

```
2022-03-22 17:08:34.388 2585-2830/? W/cronet-open: https://storage.googleapis.com/wvmedia/cenc/h264/tears/tears.mpd
2022-03-22 17:08:34.579 2585-2869/? W/cronet-open: https://storage.googleapis.com/wvmedia/cenc/h264/tears/tears_h264_main_480p_2000.mp4
2022-03-22 17:08:34.579 2585-2870/? W/cronet-open: https://storage.googleapis.com/wvmedia/cenc/h264/tears/tears_audio_eng.mp4
2022-03-22 17:08:34.759 2585-2869/? W/cronet-open: https://storage.googleapis.com/wvmedia/cenc/h264/tears/tears_h264_main_480p_2000.mp4
2022-03-22 17:08:34.760 2585-2870/? W/cronet-open: https://storage.googleapis.com/wvmedia/cenc/h264/tears/tears_audio_eng.mp4
2022-03-22 17:08:34.943 2585-2884/? W/dhds-open: https://proxy.uat.widevine.com/proxy?video_id=2015_tears&provider=widevine_test
```

As you can see in both cases the license request is initialized about 400ms after the manifest request. This seems about right - the license request should be made as soon as the encrypted content is buffered by the player, so there's a couple of round trips (manifest + content) to be made before the license request can be made.

The updates you've made to this issue so far are quite hard to follow. Please can you distil your problem down into a very easy to understand statement, and a simple set of steps we can take to reproduce. This may also mean providing links to your content and DRM license server if the problem doesn't reproduce with the content in the demo app. Please either upload it here or send to dev.exoplayer@gmail.com using a subject in the format `Issue #1234` (where `#1234` should be replaced with this issue number.) Please also update this issue to indicate you’ve done this.

**Comment by krackjack234** [comment]
Hi @icbaker  thanks for the response.

Apologies if the posts are not clear.

Our goal was to start mpd+widevine playback as soon as possible using exoplayer 2.14.1

As per the logs that we collected, we found that the time taken from DrmSessionAquired event to DrmKeysLoaded event is ~800ms-1200ms for our streams (1131ms as per the log below). This was the reason we saw a delay between the first frame loaded event and the first frame rendered event even when we have set the initial bitrate to 100kbps


```
2022-03-22 23:10:20.047 30625-30625/com.tv.v18.viola E/Profiling::: setting initial bitrate: 100000
2022-03-22 23:10:20.074 30625-30625/com.tv.v18.viola E/Profiling::: READY = true STATE = 1
2022-03-22 23:10:20.080 30625-30625/com.tv.v18.viola E/Profiling::: multisession: false
2022-03-22 23:10:20.097 30625-30625/com.tv.v18.viola E/Profiling::: READY = true STATE = 2
2022-03-22 23:10:20.098 30625-30625/com.tv.v18.viola E/Profiling: Initialising player done
2022-03-22 23:10:20.542 30625-30625/com.tv.v18.viola E/Profiling: bitrate: 100000 bytes loaded: 2848
2022-03-22 23:10:20.547 30625-30625/com.tv.v18.viola E/Profiling: drm session acquired state: 3
2022-03-22 23:10:20.923 30625-30625/com.tv.v18.viola E/Profiling: decoder init: trackType: 2 decoder: OMX.MTK.VIDEO.DECODER.AVC.secure init time sec 0
2022-03-22 23:10:20.936 30625-30625/com.tv.v18.viola E/Profiling: bitrate: 100000 bytes loaded: 108015
2022-03-22 23:10:21.046 30625-30625/com.tv.v18.viola E/Profiling: decoder init: trackType: 1 decoder: c2.android.aac.decoder init time sec 0
2022-03-22 23:10:21.064 30625-30625/com.tv.v18.viola E/Profiling: bitrate: 100000 bytes loaded: 154922
2022-03-22 23:10:21.333 30625-30625/com.tv.v18.viola E/Profiling: bitrate: 100000 bytes loaded: 150315
2022-03-22 23:10:21.350 30625-30625/com.tv.v18.viola E/Profiling: bitrate: 2300000 bytes loaded: 2854
2022-03-22 23:10:21.678 30625-30625/com.tv.v18.viola E/Profiling: drm keys loaded, time taken: 1131
2022-03-22 23:10:21.739 30625-30625/com.tv.v18.viola E/Profiling: First Frame Rendered : 1713
```

To debug this issue we checked network logs in charles and we found that the call for drm license is happening late for us.

We will check with our client and see if we can share the streams with you. Will circle back as soon as we hear from them.

One followup question @icbaker 

Why is the player making drm license call after downloading few segments? Ideally the player can make the license call as soon as the mpd file is parsed.

**Comment by krackjack234** [comment]
@icbaker  We checked exoplayer logs from 2.14.1 vanilla app with drm test urls (Widevine DASH (policy tests) - > HW secure crypto) present in the app

```
03-23 12:40:42.984 25154 25154 I ExoPlayerImpl: Init 7ad2843 [ExoPlayerLib/2.14.1] [begoniain, Redmi Note 8 Pro, Xiaomi, 29]
03-23 12:40:42.995 25154 25154 D EventLogger: state [eventTime=0.00, mediaPos=0.00, window=0, BUFFERING]
03-23 12:40:43.042 25154 25154 D EventLogger: surfaceSize [eventTime=0.05, mediaPos=0.00, window=0, 1080, 2134]
03-23 12:40:43.058 25154 25154 D EventLogger: loading [eventTime=0.06, mediaPos=0.00, window=0, period=0, true]
03-23 12:40:43.220 25154 25154 D EventLogger: timeline [eventTime=0.23, mediaPos=0.00, window=0, period=0, periodCount=1, windowCount=1, reason=SOURCE_UPDATE
03-23 12:40:43.220 25154 25154 D EventLogger:   period [734.00]
03-23 12:40:43.220 25154 25154 D EventLogger:   window [734.00, seekable=true, dynamic=false]
03-23 12:40:43.220 25154 25154 D EventLogger: ]
03-23 12:40:43.224 25154 25154 D EventLogger: videoEnabled [eventTime=0.23, mediaPos=0.00, window=0, period=0]
03-23 12:40:43.226 25154 25154 D EventLogger: audioEnabled [eventTime=0.23, mediaPos=0.00, window=0, period=0]
03-23 12:40:43.229 25154 25154 D EventLogger: tracks [eventTime=0.23, mediaPos=0.00, window=0, period=0
03-23 12:40:43.229 25154 25154 D EventLogger:   MediaCodecVideoRenderer [
03-23 12:40:43.229 25154 25154 D EventLogger:     Group:0, adaptive_supported=YES [
03-23 12:40:43.229 25154 25154 D EventLogger:       [X] Track:0, id=1, mimeType=video/avc, bitrate=772315, codecs=avc1.42c01e, res=320x142, fps=24.0, supported=YES
03-23 12:40:43.229 25154 25154 D EventLogger:       [X] Track:1, id=2, mimeType=video/avc, bitrate=1777315, codecs=avc1.4d401f, res=854x380, fps=24.0, supported=YES
03-23 12:40:43.229 25154 25154 D EventLogger:       [X] Track:2, id=3, mimeType=video/avc, bitrate=7206998, codecs=avc1.4d4028, res=1280x570, fps=24.0, supported=YES
03-23 12:40:43.229 25154 25154 D EventLogger:       [X] Track:3, id=4, mimeType=video/avc, bitrate=18320008, codecs=avc1.64002a, res=1920x856, fps=24.0, supported=YES
03-23 12:40:43.229 25154 25154 D EventLogger:     ]
03-23 12:40:43.229 25154 25154 D EventLogger:   ]
03-23 12:40:43.229 25154 25154 D EventLogger:   MediaCodecAudioRenderer [
03-23 12:40:43.229 25154 25154 D EventLogger:     Group:0, adaptive_supported=N/A [
03-23 12:40:43.230 25154 25154 D EventLogger:       [X] Track:0, id=0, mimeType=audio/mp4a-latm, bitrate=134359, codecs=mp4a.40.2, channels=2, sample_rate=44100, language=en, supported=YES
03-23 12:40:43.230 25154 25154 D EventLogger:     ]
03-23 12:40:43.230 25154 25154 D EventLogger:   ]
03-23 12:40:43.230 25154 25154 D EventLogger:   TextRenderer []
03-23 12:40:43.230 25154 25154 D EventLogger:   MetadataRenderer []
03-23 12:40:43.230 25154 25154 D EventLogger:   CameraMotionRenderer []
03-23 12:40:43.230 25154 25154 D EventLogger: ]
03-23 12:40:44.010 25154 25154 D EventLogger: downstreamFormat [eventTime=1.02, mediaPos=0.00, window=0, period=0, id=3, mimeType=video/avc, bitrate=7206998, codecs=avc1.4d4028, res=1280x570, fps=24.0]
03-23 12:40:44.088 25154 25154 D EventLogger: drmSessionAcquired [eventTime=1.09, mediaPos=0.00, window=0, period=0, state=3]
03-23 12:40:44.386 25154 25154 D EventLogger: videoDecoderInitialized [eventTime=1.39, mediaPos=0.00, window=0, period=0, OMX.MTK.VIDEO.DECODER.AVC.secure]
03-23 12:40:44.387 25154 25154 D EventLogger: videoInputFormat [eventTime=1.39, mediaPos=0.00, window=0, period=0, id=3, mimeType=video/avc, bitrate=7206998, codecs=avc1.4D4028, res=1280x570, fps=24.0]
03-23 12:40:44.389 25154 25154 D EventLogger: downstreamFormat [eventTime=1.39, mediaPos=0.00, window=0, period=0, id=0, mimeType=audio/mp4a-latm, bitrate=134359, codecs=mp4a.40.2, channels=2, sample_rate=44100, language=en]
03-23 12:40:44.420 25154 25154 D EventLogger: audioDecoderInitialized [eventTime=1.43, mediaPos=0.00, window=0, period=0, c2.android.aac.decoder]
03-23 12:40:44.421 25154 25154 D EventLogger: audioInputFormat [eventTime=1.43, mediaPos=0.00, window=0, period=0, id=0, mimeType=audio/mp4a-latm, bitrate=134359, codecs=mp4a.40.2, channels=2, sample_rate=44100, language=en]
03-23 12:40:45.326 25154 25154 D EventLogger: drmKeysLoaded [eventTime=2.33, mediaPos=0.00, window=0, period=0]
03-23 12:40:45.380 25154 25154 D EventLogger: videoSize [eventTime=2.39, mediaPos=0.00, window=0, period=0, 1280, 570]
03-23 12:40:45.382 25154 25154 D EventLogger: renderedFirstFrame [eventTime=2.39, mediaPos=0.00, window=0, period=0, Surface(name=null)/@0xf7e0214]
```
We see it took more than 1second from DrmSessionAquired (44:088) to DrmKeysLoaded (45:326). The license network call took ~400ms

![Screen Shot 2022-03-23 at 2 59 31 PM](https://user-images.githubusercontent.com/50735078/159699796-5cb87d01-250a-4950-a780-8b1ce28bc241.png)



We are looking to optimize this

Questions
=========
1. Can we ensure that the player makes the license call as soon as it downloads the manifest and does not wait for the initial segments to be downloaded?
2. If this is not possible, can the app download the drm license in parallel to the player prepare call and then pass the license to player at a later stage?

**Comment by icbaker** [comment]
> We see it took more than 1second from DrmSessionAquired (44:088) to DrmKeysLoaded (45:326). The license network call took ~400ms
>
> 1. Can we ensure that the player makes the license call as soon as it downloads the manifest and does not wait for the initial segments to be downloaded?

ExoPlayer doesn't deliberately wait for anything except DRM operations between acquiring a DRM session and loading the keys for that session. This is a guess, and you would need to add more logging to be sure, but I wonder if the extra time you're seeing is being spent in [`MediaDrm#getKeyRequest`](https://developer.android.com/reference/android/media/MediaDrm#getKeyRequest(byte[],%20byte[],%20java.lang.String,%20int,%20java.util.HashMap%3Cjava.lang.String,%20java.lang.String%3E)) and [`MediaDrm#provideKeyResponse`](https://developer.android.com/reference/android/media/MediaDrm#provideKeyResponse(byte[],%20byte[])). We've seen reports that these can be slow on some devices, e.g. https://github.com/google/ExoPlayer/issues/7390. These are framework methods provided by the Android OS, there's nothing ExoPlayer can do to make them faster, and they're necessarily on the critical path for loading DRM keys into a session (because we can't send the request to the server until we've got it from the OS, and the only way to load the keys is to pass the response back to the OS).

**Comment by icbaker** [comment]
Closing because I don't think there's any questions remaining that are clearly-stated enough to be meaningfully answered.
